### PR TITLE
types(py): Github

### PR DESCRIPTION
**Pull request by mgaeta**
I went through the files in the GitHub integration directory and added type annotations. It was pretty straightforward but I've called out some of the more complicated changes in a special commit.
```
============================== Python coverage by team ==============================
@getsentry/data                  4.81%
@getsentry/visibility            5.37%
@getsentry/workflow              5.51%
@getsentry/owners-ingest         13.52%
@getsentry/discover-n-dashboards 14.28%
@getsentry/ecosystem             28.89%
@getsentry/performance           36.53%
@getsentry/owners-native         52.07%
TOTAL                            15.76%
```

### CODE REVIEW

File: src/sentry/integrations/bitbucket/issues.py [hunk 1]
```diff
@@ -119,7 +117,7 @@ def create_issue(self, data, **kwargs):
         try:
             issue = client.create_issue(data.get("repo"), data)
         except ApiError as e:
-            self.raise_error(e)
+            raise self.raise_error(e)
```
Comment (evanh) [inline src/sentry/integrations/bitbucket/issues.py:120]: Why this change?

File: src/sentry/integrations/github/client.py [hunk 2]
```diff
@@ -1,61 +1,109 @@
...
 
-    def get_repo(self, repo):
-        return self.get(f"/repos/{repo}")
+    def get_last_commits(self, repo: str, end_sha: str) -> Sequence[JSONData]:
+        """
+        Return API request that fetches last ~30 commits
+        see https://developer.github.com/v3/repos/commits/#list-commits-on-a-repository
+        using end_sha as parameter.
+        """
+        # Explicitly typing to satisfy mypy.
```
Comment (evanh) [inline src/sentry/integrations/github/client.py:49]: I don't know if you need this comment in every function.

File: src/sentry/integrations/github/client.py [hunk 3]
```diff
@@ -78,48 +126,45 @@ def get_with_pagination(self, path, *args, **kwargs):
...
-    def get_issues(self, repo):
-        return self.get(f"/repos/{repo}/issues")
+    def get_issues(self, repo: str) -> Sequence[JSONData]:
+        issues: Sequence[JSONData] = self.get(f"/repos/{repo}/issues")
+        return issues
 
-    def search_issues(self, query):
-        return self.get("/search/issues", params={"q": query})
+    def search_issues(self, query: str) -> Mapping[str, Sequence[Mapping[str, Any]]]:
+        # Explicitly typing to satisfy mypy.Users/marcos/sentry/sentry/src/sentry/integrations/bitbucket/issues
```
Comment (evanh) [inline src/sentry/integrations/github/client.py:140]: ?

File: src/sentry/integrations/github/client.py [hunk 4]
```diff
@@ -164,15 +209,19 @@ def create_token(self):
         )
 
     def check_file(self, repo: Repository, path: str, version: str) -> str | None:
```
Comment (evanh) [inline src/sentry/integrations/github/client.py:211]: Why does this function have a return type of `str | None`? `file` is not optional.

**Review by evanh — COMMENTED** [review]
Couple comments. Also, are there functional changes here? It was hard to tell since you reorganized a couple of the files.

File: src/sentry/integrations/bitbucket/issues.py (hunk 1 above)
Comment (untitaker) [reply src/sentry/integrations/bitbucket/issues.py:120]: I believe this change was done to help with mypy's flow analysis, and it is probably safe to do assuming `raise_error` raises unconditionally, but I think it should not be done in this PR. Instead `raise_error` should return [`NoReturn`](https://mypy.readthedocs.io/en/stable/more_types.html#the-noreturn-type).

File: src/sentry/integrations/github/client.py [hunk 5]
```diff
@@ -1,61 +1,109 @@
...
+            start = link.find("<") + 1
+            end = link.find(">")
+            return link[start:end]
 
-    def get_last_commits(self, repo, end_sha):
-        # return api request that fetches last ~30 commits
-        # see https://developer.github.com/v3/repos/commits/#list-commits-on-a-repository
-        # using end_sha as parameter
-        return self.get_cached(f"/repos/{repo}/commits", params={"sha": end_sha})
+    return None
```
Comment (untitaker) [inline src/sentry/integrations/github/client.py:31]: Why move this function out in the same PR? This could've easily stayed inline.

File: src/sentry/integrations/github/integration.py [hunk 6]
```diff
@@ -79,32 +81,37 @@
...
 
+        for f in files["items"]:
+            if f["name"] == "CODEOWNERS":
+                filepath = f["path"]
+                html_url = f["html_url"]
+                try:
+                    contents = self.get_client().get_file(repo.name, filepath)
+                except ApiError:
+                    return None
+                return {"filepath": filepath, "html_url": html_url, "raw": contents}
```
Comment (untitaker) [inline src/sentry/integrations/github/integration.py:111]: Why restructure this code?

File: src/sentry/integrations/github/integration.py [hunk 7]
```diff
@@ -117,15 +124,16 @@ def get_repositories(self, query=None):
             {"name": i["name"], "identifier": i["full_name"]} for i in response.get("items", [])
         ]
 
-    def search_issues(self, query):
+    def search_issues(self, query: str) -> Mapping[str, Sequence[Mapping[str, Any]]]:
+        # TODO MARCOS FIRST
```
Comment (untitaker) [inline src/sentry/integrations/github/integration.py:128]: ?

File: src/sentry/integrations/github/integration.py [hunk 8]
```diff
@@ -135,22 +143,25 @@ def get_unmigratable_repositories(self):
...
-            message = API_ERRORS.get(exc.code)
-            if exc.code == 404 and re.search(r"/repos/.*/(compare|commits)", exc.url):
-                message += f" Please also confirm that the commits associated with the following URL have been pushed to GitHub: {exc.url}"
-
-            if message is None:
-                message = exc.json.get("message", "unknown error") if exc.json else "unknown error"
-            return f"Error Communicating with GitHub (HTTP {exc.code}): {message}"
-        else:
+    def message_from_error(self, exc: Exception) -> str:
+        if not isinstance(exc, ApiError):
```
Comment (untitaker) [inline src/sentry/integrations/github/integration.py:150]: why rewrite this?

File: src/sentry/integrations/github/webhook.py [hunk 9]
```diff
@@ -298,28 +328,47 @@ def _handle(self, integration, event, organization, repo, host=None):
...
+        After a pull request is merged, the attribute changes depending on how
+        the pull request was merged:
+        - If the pull request was merged as a merge commit, the attribute
+         represents the SHA of the merge commit.
+        - If the pull request was merged via a squash, the attribute
+         represents the SHA of the squashed commit on the base branch.
+        - If the pull request was rebased, the attribute represents the commit
+         that the base branch was updated to.
+        https://developer.github.com/v3/pulls/#get-a-single-pull-request
+        """
```
Comment (untitaker) [inline src/sentry/integrations/github/webhook.py:371]: I don't think this code comment is of any use to a caller of `_handle`, so I don't think it's an appropriate docstring.

**Review by untitaker — COMMENTED** [review]
I've already said this before, but when adding the typing team for review, I think it would be best to constrain your changes to literally just adding the type annotations and nothing else. This PR touches a lot of files at once, adds subtle behavioral changes here and there (some of which don't have to be in this PR), is therefore _unnecessarily risky_ to review and deploy, and IMO the typing team really cannot and should not attempt to review this code like that. I reviewed some parts of it regardless to point out what I mean.

Some changes, like assigning temp variables or using `[a, *b, c]` instead of `[a] + b + [c]` are reasonable to have within such typing PRs, but also carry risk when deployed in bulk. I would like to see this PR split up in addition to not having any sort of adjacent refactoring in it.

File: src/sentry/integrations/bitbucket/issues.py (hunk 1 above)
Comment (mgaeta) [reply src/sentry/integrations/bitbucket/issues.py:120]: `NoReturn` is pretty cool. I'll split this off into a separate PR. It'll be tricky to get working because `raise_error()` isn't implemented in the mixin. I think I can get around this with a shared interface.

File: src/sentry/integrations/github/client.py (hunk 2 above)
Comment (mgaeta) [reply src/sentry/integrations/github/client.py:49]: You're right, I'll stop leaving the comment when the file is in `mypy.ini`.

File: src/sentry/integrations/github/client.py (hunk 3 above)
Comment (mgaeta) [reply src/sentry/integrations/github/client.py:140]: This looks like a bad copy/paste. I've been having really bad input delays on my machine recently :/

File: src/sentry/integrations/github/client.py (hunk 4 above)
Comment (mgaeta) [reply src/sentry/integrations/github/client.py:211]: Ok this is weird so I'll leave a comment in the file. _GitLab's_ `check_file()` returns an optional string so in order for `RepositoryMixin` to type-check they're the same method signature.

File: src/sentry/integrations/bitbucket/issues.py (hunk 1 above)
Comment (mgaeta) [reply src/sentry/integrations/bitbucket/issues.py:120]: Ok I had to give up. I can't commit to these files at all without satisfying `pyright` and it's getting `reportUnboundVariable` errors here.

File: src/sentry/integrations/github/webhook.py (hunk 9 above)
Comment (mgaeta) [reply src/sentry/integrations/github/webhook.py:371]: I agree but when I was working on the revert I realized this isn't being detected as docstring it's just a multi-line string which I think is acceptable comment syntax.

**Comment by mgaeta** [comment]
I just reverted a lot of the unnecessary changes and I'll put them up in a separate PR.
There shouldn't be any functional changes but fixing the `reportUnboundVariable` errors could technically be a change.

File: src/sentry/integrations/bitbucket/issues.py (hunk 1 above)
Comment (untitaker) [reply src/sentry/integrations/bitbucket/issues.py:125]: Just tried it out locally because I couldn't believe pyright would let you down like that, but... yeah. I think this is fine then. It certainly beats `return raise_error()` which might accidentally noop.

File: src/sentry/integrations/github/webhook.py (hunk 9 above)
Comment (untitaker) [reply src/sentry/integrations/github/webhook.py:372]: I've not seen people use multi-line strings like that but I don't have a strong opinion on it.

**Review by untitaker — APPROVED** [review]
